rs/xnet/payload_builder/src/lib.rs

```diff
@@ -879,11 +883,24 @@ impl XNetPayloadBuilderImpl {
 ///
 /// In order to prevent mutual stalling, only applies to incoming NNS
 /// streams; and to `Application`-subnet-to-`System`-subnet streams.
-pub fn get_msg_limit(subnet_id: SubnetId, state: &ReplicatedState) -> Option<usize> {
+pub fn get_msg_limit(
+    subnet_id: SubnetId,
+    state: &ReplicatedState,
+    expected: Option<&ExpectedIndices>,
+    slice_begin: StreamIndex,
+) -> Option<usize> {
+    let get_limit = || -> Option<usize> {
+        const MAX_SIGNALS_STREAM_COUNT: usize = 50_000;
+        let delta = expected.map_or(0, |expected| {
+            (expected.message_index.get() - slice_begin.get()) as usize
+        });
+        Some(MAX_SIGNALS_STREAM_COUNT - delta)
```

Comment:
I'd rather put this as in my first comment on the PR: `stream_begin + 50k - expected.message_index()`.